File: frontend/src/api/market.ts
// 시장 데이터 API

import {
  api,
  ApiResponse,
  ApiClientError,
  ensureApiSuccess,
  toApiClientError,
  unwrapApiResponse,
} from './config';
import type {
  FearGreedIndex,
  MarketPrice,
  OHLCV,
  SRLevel,
  TradeReportData,
  VPVRData,
  VPVRSourceData,
} from '../types';
import { isOHLCV } from '../utils/ohlcv';

interface OHLCVResponse {
  data: OHLCV[];
  source: string;
  count: number;
}

export async function getMarketPrices(): Promise<Record<string, MarketPrice>> {
  try {
    const res = await api.get<ApiResponse<Record<string, MarketPrice>>>('/market/prices');
    return unwrapApiResponse(res, 'Failed to load market prices.');
  } catch (error: unknown) {
    throw toApiClientError(error, 'Failed to load market prices.');
  }
}

export async function getFearGreedIndex(): Promise<FearGreedIndex> {
  try {
    const res = await api.get<ApiResponse<FearGreedIndex>>('/market/fear-greed');
    return unwrapApiResponse(res, 'Failed to load the fear and greed index.');
  } catch (error: unknown) {
    throw toApiClientError(error, 'Failed to load the fear and greed index.');
  }
}

export async function getTimeframes(coin: string): Promise<{
  all: string[];
  binance: string[];
  csv: string[];
}> {
  try {
    const res = await api.get<ApiResponse<{ all: string[]; binance: string[]; csv: string[] }>>(
      `/timeframes/${coin}`
    );
    return unwrapApiResponse(res, 'Failed to load available timeframes.');
  } catch (error: unknown) {
    throw toApiClientError(error, 'Failed to load available timeframes.');
  }
}

export async function getOHLCV(
  coin: string,
  interval: string,
  useCsv: boolean = false,
  limit: number = 3000,
  endTime?: number,
): Promise<OHLCVResponse> {
  try {
    const res = await api.get<{
      success: boolean;
      data?: unknown;
      source: string;
      count: number;
      error?: string;
      error_code?: string | null;
      details?: unknown;
    }>(`/ohlcv/${coin}/${interval}`, {
      params: { use_csv: useCsv, limit, end_time: endTime },
      timeout: 30000,
    });
    const payload = ensureApiSuccess(res, 'Failed to load OHLCV data.');

    if (!Array.isArray(payload.data) || !payload.data.every(isOHLCV)) {
      throw new ApiClientError('OHLCV response contains invalid candle data.');
    }

    return {
      data: payload.data,
      source: payload.source,
      count: payload.count,
    };
  } catch (error: unknown) {
    throw toApiClientError(error, 'Failed to load OHLCV data.');
  }
}

export async function getTradeReport(
  coin: string,
  interval: string,
  options: {
    limit?: number;
    end_time?: number;
    as_of?: number;
    profile_candles?: number;
  },
): Promise<TradeReportData> {
  try {
    const res = await api.get<ApiResponse<TradeReportData>>(
      `/indicators/trade-report/${coin}/${interval}`,
      { params: options, timeout: 30000 },
    );
    const payload = unwrapApiResponse(res, 'Failed to load the trade report.');
    if (!Array.isArray(payload.candles) || !payload.candles.every(isOHLCV)) {
      throw new ApiClientError('Trade report response contains invalid candle data.');
    }
    return payload;
  } catch (error: unknown) {
    throw toApiClientError(error, 'Failed to load the trade report.');
  }
}

export async function getVPVRSource(
  coin: string,
  interval: string,
  candles?: number
): Promise<VPVRSourceData> {
  try {
    const res = await api.get<ApiResponse<VPVRSourceData>>(
      `/indicators/vpvr-source/${coin}/${interval}`,
      { params: { candles }, timeout: 20000 }
    );
    return unwrapApiResponse(res, 'Failed to load Binance VPVR source data.');
  } catch (error: unknown) {
    throw toApiClientError(error, 'Failed to load Binance VPVR source data.');
  }
}

export async function getVPVR(
  coin: string,
  interval: string,
  options?: { candles?: number; bin_count?: number; price_range?: number }
): Promise<VPVRData> {
  try {
    const res = await api.get<ApiResponse<VPVRData>>(
      `/indicators/vpvr/${coin}/${interval}`,
      { params: options, timeout: 20000 }
    );
    return unwrapApiResponse(res, 'Failed to calculate the Binance volume profile.');
  } catch (error: unknown) {
    throw toApiClientError(error, 'Failed to calculate the Binance volume profile.');
  }
}

export async function getSupportResistance(
  coin: string,
  interval: string,
  options?: {
    lookback?: number;
    tolerance_pct?: number;
    min_touches?: number;
    show_pivots?: boolean;
    htf_option?: string;
  }
): Promise<SRLevel[]> {
  try {
    const res = await api.get<ApiResponse<SRLevel[]>>(`/support-resistance/${coin}/${interval}`, {
      params: options,
    });
    return unwrapApiResponse(res, 'Failed to load support and resistance levels.');
  } catch (error: unknown) {
    throw toApiClientError(error, 'Failed to load support and resistance levels.');
  }
}

File: backend/routes/market.py
# routes/market.py
"""시장 데이터 API 엔드포인트"""

from fastapi import APIRouter
import pandas as pd

from utils.data_service import get_market_prices, get_fear_and_greed_index, discover_timeframes, fetch_live_data
from utils.data_loader import load_data_for_analysis
from utils.decorators import handle_api_errors
from core.support_resistance import (
    build_sr_levels_from_swings,
    compute_daily_pivots,
    compute_htf_sr_levels,
)

router = APIRouter(prefix="/api", tags=["market"])


@router.get("/market/prices")
@handle_api_errors()
async def api_market_prices():
    """Get current market prices for major coins"""
    prices = get_market_prices()
    if prices is None:
        return {"success": False, "data": None, "error": "Failed to fetch prices"}
    
    result = {}
    for symbol, data in prices.items():
        coin = symbol.replace("/USDT", "")
        result[coin] = {
            "last": data.get("last", 0),
            "percentage": data.get("percentage", 0),
            "high": data.get("high", 0),
            "low": data.get("low", 0),
            "volume": data.get("quoteVolume", 0),
        }
    return {"success": True, "data": result}


@router.get("/market/fear-greed")
@handle_api_errors()
async def api_fear_greed():
    """Get Fear & Greed Index"""
    fng = get_fear_and_greed_index()
    if fng is None:
        return {"success": False, "data": None}
    return {"success": True, "data": fng}


@router.get("/timeframes/{coin}")
@handle_api_errors()
async def api_timeframes(coin: str):
    """Get available timeframes for a coin"""
    all_tfs, binance_tfs, csv_tfs = discover_timeframes(coin)
    return {
        "success": True,
        "data": {
            "all": all_tfs,
            "binance": list(binance_tfs),
            "csv": csv_tfs,
        }
    }


@router.get("/ohlcv/{coin}/{interval}")
@handle_api_errors()
async def api_ohlcv(coin: str, interval: str, use_csv: bool = False, limit: int = 3000):
    """Get OHLCV data for a coin and interval"""
    # Load data using common loader
    df, source = load_data_for_analysis(coin, interval, use_csv, total_candles=limit)
    
    # Convert to JSON-serializable format
    df["open_dt"] = df["open_dt"].astype(str)
    data = df.to_dict(orient="records")
    
    return {
        "success": True,
        "data": data,
        "source": source,
        "count": len(data),
    }


@router.get("/support-resistance/{coin}/{interval}")
@handle_api_errors()
async def api_support_resistance(
    coin: str,
    interval: str,
    lookback: int = 3,
    tolerance_pct: float = 0.3,
    min_touches: int = 3,
    show_pivots: bool = False,
    htf_option: str = "none",
):
    """Calculate support/resistance levels"""
    df = fetch_live_data(f"{coin}/USDT", interval, total_candles=1000)
    if df is None or df.empty:
        return {"success": False, "error": "Failed to load data"}
    
    frames = []
    
    # Swing-based SR levels
    sr_main = build_sr_levels_from_swings(
        df,
        lookback=lookback,
        tolerance_pct=tolerance_pct / 100.0,
        min_touches=min_touches,
        timeframe_label=interval,
    )
    if not sr_main.empty:
        frames.append(sr_main)
    
    # Daily pivots
    if show_pivots:
        pivots = compute_daily_pivots(df, last_n=1)
        if not pivots.empty:
            frames.append(pivots)
    
    # HTF levels
    if htf_option != "none":
        rule = "4H" if htf_option == "4H" else "1D"
        htf_levels = compute_htf_sr_levels(
            df,
            rule=rule,
            lookback=lookback,
            tolerance_pct=tolerance_pct / 100.0,
            min_touches=min_touches,
        )
        if not htf_levels.empty:
            frames.append(htf_levels)
    
    if frames:
        all_levels = pd.concat(frames, ignore_index=True)
        return {"success": True, "data": all_levels.to_dict(orient="records")}
    
    return {"success": True, "data": []}
